server/core/region.go

```diff
@@ -475,15 +475,67 @@ func (r *RegionsInfo) GetFollower(storeID uint64, regionID uint64) *RegionInfo {
 
 // ScanRange scans region with start key, until number greater than limit.
 func (r *RegionsInfo) ScanRange(startKey []byte, limit int) []*RegionInfo {
-	metaRegions := r.tree.scanRange(startKey, limit)
-	res := make([]*RegionInfo, 0, len(metaRegions))
-	for _, m := range metaRegions {
-		region := r.GetRegion(m.region.GetId())
-		res = append(res, region)
-	}
+	res := make([]*RegionInfo, 0, limit)
+	r.tree.scanRange(startKey, func(region *metapb.Region) bool {
+		res = append(res, r.GetRegion(region.GetId()))
+		return len(res) < limit
+	})
 	return res
 }
 
+// RegionStats records a list of regions' statistics and distribution status.
+type RegionStats struct {
+	Count            int              `json:"count"`
+	EmptyCount       int              `json:"empty_count"`
+	StorageSize      int64            `json:"storage_size"`
+	StoreLeaderCount map[uint64]int   `json:"store_leader_count"`
+	StorePeerCount   map[uint64]int   `json:"store_peer_count"`
+	StoreLeaderSize  map[uint64]int64 `json:"store_leader_size"`
+	StorePeerSize    map[uint64]int64 `json:"store_peer_size"`
+}
+
+func newRegionStats() *RegionStats {
+	return &RegionStats{
+		StoreLeaderCount: make(map[uint64]int),
+		StorePeerCount:   make(map[uint64]int),
+		StoreLeaderSize:  make(map[uint64]int64),
+		StorePeerSize:    make(map[uint64]int64),
+	}
+}
+
+// Observe adds a region's statistics into RegionStats.
+func (s *RegionStats) Observe(r *RegionInfo) {
+	s.Count++
+	if r.ApproximateSize <= 1 {
```

Comment:
why 1 here?

Reply:
Regions that have approximate size less than 1MB is record as 1.
See: https://github.com/pingcap/pd/blob/96db717124a5bd980e8a244af2db8952e04159d6/server/grpc_service.go#L325

Reply:
oh, I think we should the size should be more accurate, maybe KB is better.

Reply:
As @zhangjinpeng1987 suggested, tikv's calculation of size is not accurate in the first place, using MB will reduce inaccuracy relatively.

Reply:
please define a const var

Reply:
please define a const var